### PR TITLE
Onfleet node

### PR DESCRIPTION
**Pull request by 0xsemaj**
### Preparation

- [x]  Set up your editor for code formatting (indentation, new lines, linting). If you use Visual Studio Code, you can use the TSLint extension for linting.
- [x]  Get credentials (e.g., Client ID, Client Secret, API key, user login, user password, website URL) for the service you are building a node for.

### Development

- [x] Open a pull request as early as possible with WIP in the pull request title. (We're doing this in one big PR 😆 )
- [x] If you are creating a node requested by a community member, make sure to comment on the feature request in the community forum (opens new window).
- [x] Ensure complementary operations to each resource (e.g., `create`, `delete`) have been added.
- [x] Ensure the node works with multiple items via one input.
- [x] Ensure the parameters have the correct type.
- [x] Mind the defaults: if the service has a default as `true`, keep it as `true`. Changing default values can break the existing workflows of the users.
- [x] Check if the node disposes of everything properly, in particular, if connections were properly closed.
- [x] Check your code using Nodelinter to ensure a clean lint before submitting your pull request

### Testing

- [x] Test `create` and `update` operations with all fields/operations.
- [x] Test the `continueOnFail` option with a Function node. (For example, a Widget node has a GET operation that takes a widgetId and returns information on the widget. To test that the workflow continues on fail, set the Widget node to continue on fail, create a Function node, return a valid and an invalid widgetId, connect the Function node to Widget node, and run the workflow. The Widget node should show two items: one with information on the widget and another one with the error from having passed an invalid ID.)

### Code formatting
- [x] Ensure the branch lints cleanly by running npm run lint.
- [x] Ensure the indentation is correct. Check this in the `editorconfig`.
- [x] Ensure there are no extra spaces. Check this in the `editorconfig`.
- [x] Code comment dividers inside if-branches.
- [x] Use "create/delete" verbs for operations, except for tags, where you should use "add/remove". (We still opt to use `create` in several cases to be uniform with our API documentation)

### Errors and Outputs
- [x] Ensure empty API responses return { success: true }.
- [x] Ensure the error responses are handled and displayed correctly (e.g., malformed requests, requests with invalid credentials) and use the current format. You can check this by making failing requests to the API.
- [x] Check if the response can be simplified and add a simplify function (e.g., SecurityScorecard node). (Not needed in our case)
- [x] Ensure the response from `Create` is consistent with `Get`.
- [x] Ensure the response from `Get All` is consistent with `Get`.

### Presentation
- [x] Ensure the primary menu contains only required parameters.
- [x] Ensure a JSON object is not shown in a single column in Table view. (Nested JSON)
- [x] Make sure all `GetAll` operations have the fields return and limit.
- [x] Ensure the property subtitle is set.
- [x] Make sure the pagination (if any) is working correctly. Set Limit 1.

### Writing
- [x] Ensure all descriptions are correct and end with a period.
- [x] Ensure that most descriptions exist, excluding redundant ones.
- [x] Ensure IDs in `displayNames` are capitalized (i.e.: "IDs", not "ids" or "Ids").
- [x] Ensure that IDs, if multiple, have descriptive qualifiers.
- [x] Ensure the name property in `description` in the node class is written in camelCase.
- [x] Ensure the file name and the Class name are identical.

### Branding
- [x] Ensure the name of the service is written correctly (e.g., "GitHub" not "Github"). If the node is a trigger node, ensure it is named as such, by adding "Trigger" after the service name (e.g., "Trello Trigger").
- [x] Ensure the logo is either a PNG or SVG, ideally the latter. Vecta is a good website to find SVGs of different applications.
- [x] If the logo is an SVG, ensure the canvas is a perfect square. If the logo is PNG, ensure it is 60x60 pixels and compressed.
- [x] Ensure the border color of the node matches the branding of the service.

### Nice-to-haves (optional)
- [x] Add handler for `continueOnFail`. This feature is included in some of the newest nodes (e.g Lemlist node) to continue the workflow even if the node's execution fails.
- [x] Remove required: `false` and `description: ''` in the node descriptions (e.g., Lemlist node).
- [x] At call site, specify first `body` and then `qs`.
- [x] At call site, prepend the endpoint with slash `/` (e.g., "/campaign").

### CODE REVIEW

**Comment by jamesliupenn** [comment]
@ivov hi there, do you know what's the typical lead time to get a new node added? We're anticipating this release and hope to get an estimate of when this will be live.

**Comment by janober** [comment]
Thanks a lot for the contribution @jamesliupenn. The node is one of the next ones that will get reviewed. So should not be much longer.

**Comment by jamesliupenn** [comment]
Thanks! I realized we missed a few operations so we'll be making some updates tomorrow

File: packages/nodes-base/nodes/Onfleet/descriptions/TeamDescription.ts [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+const managersField = {
+	displayName: 'Administrators (JSON)',
+	name: 'managers',
+	type: 'json',
+	default: '[]',
+	description: 'An array of managing administrator IDs.',
+} as INodeProperties;
+
+const hubField = {
+	displayName: 'Hub',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/TeamDescription.ts:71]: Hub ID. Can you load all the available Hubs?

File: packages/nodes-base/nodes/Onfleet/descriptions/TeamDescription.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+const enableSelfAssignmentField = {
+	displayName: 'Self assignment',
+	name: 'enableSelfAssignment',
+	type: 'boolean',
+	default: false,
+	description: 'This toggles Team Self-Assignment that allows Drivers to Self Assign Tasks that are in the Team unassigned container.',
+} as INodeProperties;
+
+export const teamFields = [
+	{
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/TeamDescription.ts:87]: Team ID

File: packages/nodes-base/nodes/Onfleet/descriptions/TeamDescription.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+		displayOptions: {
+			show: {
+				resource: [ 'teams' ],
+				operation: [ 'create' ],
+			},
+		},
+		required: true,
+	},
+	{
+		displayName: 'Additional fields',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/TeamDescription.ts:136]: Additional Fields

File: packages/nodes-base/nodes/Onfleet/descriptions/TeamDescription.ts [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+				operation: [ 'create' ],
+			},
+		},
+		options: [
+			hubField,
+			enableSelfAssignmentField,
+		],
+	},
+	{
+		displayName: 'Additional fields',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/TeamDescription.ts:153]: Additional Fields

File: packages/nodes-base/nodes/Onfleet/descriptions/WebhookDescription.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
+import {
+	INodeProperties
+} from 'n8n-workflow';
+import { webhookMapping } from '../WebhookMapping';
+
+export const webhookOperations = [
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WebhookDescription.ts:6]: Why is this resource needed when the trigger is already handling these endpoints?

Comment (RicardoE105) [reply packages/nodes-base/nodes/Onfleet/descriptions/WebhookDescription.ts:6]: I would remove it complely

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+				resource: [ 'workers' ],
+			},
+		},
+		options: [
+			{
+				name: 'Create',
+				value: 'create',
+				description: 'Create a new Onfleet worker.',
+			},
+			{
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:21]: Delete

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+				name: 'Remove',
+				value: 'delete',
+				description: 'Remove an Onfleet worker.',
+			},
+			{
+				name: 'List',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:27]: Get All

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+				value: 'getAll',
+				description: 'List all Onfleet workers.',
+			},
+			{
+				name: 'List workers by location',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:32]: Can we merge this operation and the getAll operation?

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 9]
```diff
@@ -0,0 +1,529 @@
...
+				value: 'getAllByLocation',
+				description: 'List all Onfleet workers who are currently within a centain target area.',
+			},
+			{
+				name: 'Get',
+				value: 'get',
+				description: 'Get a specific Onfleet worker.',
+			},
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:40]: Sort alphabetically

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 10]
```diff
@@ -0,0 +1,529 @@
...
+	name: 'type',
+	type: 'options',
+	options: [
+		{
+			name: 'Car',
+			value: 'CAR',
+		},
+		{
+			name: 'Motorcycle',
+			value: 'MOTORCYCLE',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:113]: Sort alphabetically

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 11]
```diff
@@ -0,0 +1,529 @@
...
+const vehicleColorField = {
+	displayName: 'Color',
+	name: 'color',
+	type: 'string',
+	default: '',
+	description: 'The vehicle\'s color.',
+} as INodeProperties;
+
+const teamsField = {
+	displayName: 'Teams (JSON)',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:153]: No need to do a JSON. Use string with mutlipleValues set to true

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 12]
```diff
@@ -0,0 +1,529 @@
...
+
+const statesFilterField = {
+	displayName: 'States',
+	name: 'states',
+	type: 'multiOptions',
+	options: [
+		{
+			name: 'Off-duty',
+			value: 0,
+		},
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:176]: Sort alphabetically

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 13]
```diff
@@ -0,0 +1,529 @@
...
+	name: 'phones',
+	type: 'string',
+	default: '',
+	description: 'A comma-separated list of workers\' phone numbers.',
+} as INodeProperties;
+
+const filterField = {
+	displayName: 'List of fields to return',
+	name: 'filter',
+	type: 'string',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:201]: Can we load all the possible fields so they do not have to type them in?

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 14]
```diff
@@ -0,0 +1,529 @@
...
+	},
+	default: 1000,
+	description: 'The length in meters of the radius of the spherical area in which to look for workers. Defaults to 1000 if missing. Maximum value is 10000.',
+} as INodeProperties;
+
+const scheduleDateField = {
+	displayName: 'Date',
+	name: 'date',
+	type: 'dateTime',
+	default: Date.now(),
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:244]: Set default to empty

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 15]
```diff
@@ -0,0 +1,529 @@
...
+	description: 'Schedule\'s date.',
+} as INodeProperties;
+
+const scheduleTimezoneField = {
+	displayName: 'Timezone',
+	name: 'timezone',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:250]: Load all timezones from moment.

			async getTimezones(
				this: ILoadOptionsFunctions,
			): Promise<INodePropertyOptions[]> {
				const returnData: INodePropertyOptions[] = [];
				for (const timezone of moment.tz.names()) {
					const timezoneName = timezone;
					const timezoneId = timezone;
					returnData.push({
						name: timezoneName,
						value: timezoneId,
					});
				}
				return returnData;
			},

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 16]
```diff
@@ -0,0 +1,529 @@
...
+	type: 'string',
+	default: '',
+	description: 'A valid timezone.',
+} as INodeProperties;
+
+const scheduleStartField = {
+	displayName: 'Start',
+	name: 'start',
+	type: 'dateTime',
+	default: Date.now(),
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:260]: Set default to ''

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 17]
```diff
@@ -0,0 +1,529 @@
...
+	description: 'Start time.',
+} as INodeProperties;
+
+const scheduleEndField = {
+	displayName: 'End',
+	name: 'end',
+	type: 'dateTime',
+	default: Date.now(),
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:268]: Set default to ''

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 18]
```diff
@@ -0,0 +1,529 @@
...
+	description: 'End time.',
+} as INodeProperties;
+
+export const workerFields = [
+	{
+		displayName: 'ID',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:274]: Worker ID

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts (hunk 7 above)
Comment (RicardoE105) [reply packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:27]: Pagination is missing

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts (hunk 8 above)
Comment (RicardoE105) [reply packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:32]: As long as they report the same data I think they can be merged.

Comment (RicardoE105) [reply packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:32]: Pagination is missing

Comment (RicardoE105) [reply packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:32]: The collection needs to be called filters.

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 19]
```diff
@@ -0,0 +1,529 @@
...
+				vehicle: [ true ],
+			},
+		},
+		options: [
+			vehicleDescriptionField,
+			vehicleLicensePlateField,
+			vehicleColorField,
+		],
+	},
+	{
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:420]: Title case

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 20]
```diff
@@ -0,0 +1,529 @@
...
+				operation: [ 'create' ],
+			},
+		},
+		options: [
+			capacityField,
+			displayNameField,
+		],
+	},
+	{
+		displayName: 'Additional fields',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:438]: Update Fields

File: packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts [hunk 21]
```diff
@@ -0,0 +1,529 @@
...
+			filterField,
+			teamsFilterField,
+			statesFilterField,
+			phonesFilterField,
+		],
+	},
+	{
+		displayName: 'Additional fields',
+		name: 'additionalFields',
+		type: 'collection',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:478]: The analytics fields need to be optional.

Comment (RicardoE105) [reply packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:478]: Rename the collection to Options.

**Review by RicardoE105 — COMMENTED** [review]
Got reviewed. Let me know if you have questions about any comments. @jamesliupenn

**Comment by jamesliupenn** [comment]
We'll be prioritizing this in the ongoing sprint, thanks for the feedback @RicardoE105, we'll keep you posted once we finish developing it internally.

File: packages/nodes-base/nodes/Onfleet/descriptions/ContainerDescription.ts [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+		name: 'operation',
+		type: 'options',
+		displayOptions: {
+			show: {
+				resource: [ 'containers' ],
+			},
+		},
+		options: [
+			{
+				name: 'Insert tasks (or append)',
```
[the remark this answers is not in the document]
Comment (YokySantiago) [reply packages/nodes-base/nodes/Onfleet/descriptions/ContainerDescription.ts:17]: Hey @RicardoE105 !
In this case we don't have container information. This is the meaning of container for us "Containers are an abstraction which describes task assignment. A container is an ordered list of tasks. Organizations, teams and workers all correspond to containers as they can all be assigned tasks." So, when we get information about a container, we need to select the type (Organization, Team, Worker) in order to get the task assignment for the container selected.
So, We would like to keep it all together as a resource

File: packages/nodes-base/nodes/Onfleet/OnfleetTrigger.node.ts [hunk 23]
```diff
@@ -0,0 +1,224 @@
...
+	};
+
+	// @ts-ignore (because of request)
+	webhookMethods = {
+		default: {
+			async checkExists(this: IHookFunctions): Promise<boolean> {
+				const webhookData = this.getWorkflowStaticData('node') as IDataObject;
+				const credentials = await this.getCredentials('onfleetApi') as ICredentialDataDecryptedObject;
+				const event = this.getNodeParameter('event', 0) as string;
+				const { name = '' } = this.getNodeParameter('additionalFields', 0) as IDataObject;
```
[the remark this answers is not in the document]
Comment (YokySantiago) [reply packages/nodes-base/nodes/Onfleet/OnfleetTrigger.node.ts:67]: Hey Ricardo! Well, this field is part of the API offering and we'd like to keep them for uniformity

File: packages/nodes-base/nodes/Onfleet/descriptions/OrganizationDescription.ts [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+		},
+		options: [
+			{
+				name: 'Get details',
+				value: 'get',
+				description: 'Retrieve your own organization\'s details.',
+			},
+			{
+				name: 'Get delegatee details',
+				value: 'getDelegatee',
```
[the remark this answers is not in the document]
Comment (jamesliupenn) [reply packages/nodes-base/nodes/Onfleet/descriptions/OrganizationDescription.ts:23]: We would like things to still relate to our API docs for better support, especially on the language used. In this case, delegatee has always been used and the change in terminology would simply confuse our users.

File: packages/nodes-base/nodes/Onfleet/OnfleetTrigger.node.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (jamesliupenn) [reply packages/nodes-base/nodes/Onfleet/OnfleetTrigger.node.ts:67]: I would have to agree on keeping the name, many of our customers use the name as an indicator of what types of webhooks it is and where it is utilized. Using n8n-url would simply be too generic

File: packages/nodes-base/nodes/Onfleet/descriptions/TaskDescription.ts [hunk 25]
```diff
@@ -0,0 +1,380 @@
...
+			},
+		},
+		options: [
+			{
+				name: 'Create',
+				value: 'create',
+				description: 'Create a new Onfleet task.',
+			},
+			{
+				name: 'Create multiple tasks',
```
[the remark this answers is not in the document]
Comment (jamesliupenn) [reply packages/nodes-base/nodes/Onfleet/descriptions/TaskDescription.ts:22]: Batch create and create are two different operations in our API, they should both be supported in cases where some users need to bulk upload tasks versus when someone stream processes task creation

File: packages/nodes-base/nodes/Onfleet/descriptions/TaskDescription.ts [hunk 26]
```diff
@@ -0,0 +1,380 @@
...
+				value: 'complete',
+				description: 'Force-complete a started Onfleet task.',
+			},
+			{
+				name: 'Remove',
+				value: 'delete',
+				description: 'Remove an Onfleet task.',
+			},
+			{
+				name: 'List',
```
[the remark this answers is not in the document]
Comment (jamesliupenn) [reply packages/nodes-base/nodes/Onfleet/descriptions/TaskDescription.ts:42]: _from_ is not optional when listing all tasks, it is actually required https://docs.onfleet.com/reference#list-tasks

File: packages/nodes-base/nodes/Onfleet/descriptions/WebhookDescription.ts (hunk 5 above)
Comment (YokySantiago) [reply packages/nodes-base/nodes/Onfleet/descriptions/WebhookDescription.ts:6]: Well, the trigger node is a listener, however, we'd like to run webhook operations for other apps from n8n nodes.  i.e: 

- We want to be able to create an Onfleet webhook but n8n is not the listener
- We want to delete an Onfleet webhook however n8n is not the listener
- We want to list created Onfleet webhooks and they aren't necessary of n8n

In addition, this is part of the API offering and we'd like to keep them for uniformity.

File: packages/nodes-base/nodes/Onfleet/descriptions/TaskDescription.ts [hunk 27]
```diff
@@ -0,0 +1,380 @@
...
+					{
+						name: 'Completed',
+						value: 3,
+					},
+				],
+				default: null,
+				description: 'The state of the tasks.',
+			},
+			{
+				displayName: 'LastId',
```
[the remark this answers is not in the document]
Comment (jamesliupenn) [reply packages/nodes-base/nodes/Onfleet/descriptions/TaskDescription.ts:234]: Yes this is needed for pagination in `/tasks/all` endpoint

File: packages/nodes-base/nodes/Onfleet/descriptions/WebhookDescription.ts (hunk 5 above)
Comment (jamesliupenn) [reply packages/nodes-base/nodes/Onfleet/descriptions/WebhookDescription.ts:6]: There are some use cases where n8n is used as middleware and not necessarily the webhook server. Hence we would like to keep the current setup so the webhook resource is handling endpoints outside of the trigger node.

**Comment by jamesliupenn** [comment]
@RicardoE105 this is ready for another round of review as we have resolved the issues you mentioned. Please keep us posted👍👍

**Comment by RicardoE105** [comment]
@jamesliupenn perfect. Will review it again as soon as I can.

**Comment by RicardoE105** [comment]
@jamesliupenn had a chance to review again. This time I did a deeper review than last time and found the following. Can you please address these issues? Thanks.

Onfleet review

Admin resource

Admin:getAll

- [ ] Admin:getall missing pagination

Admin:delete

- [ ] Return { success: true } when the operation is successfull

Task resource

- [ ] Do only task:create and do the batching behind the scenes. Batching if available should be the implementation to use.

Task:create 

- [ ] Every field that is not required should be under a collection. I was able to create a task only with the destination address, which means, that recipients it’s not required
- [ ] When setting the address you have a collection called “additional fields” that is not necessary. Add them at the same level of the destination address and make them not required. Do the same with `Recipient`
- [ ] Rename additional task fields to Additional Fields

Task:clone
- [ ] If the operation does not return anything when it ran successfully return { success: true }
- [ ] Rename collection options to additional fields
- [ ] Move what is inside the collection override to the same level the options parameters are

Task:complete
- [ ] If this operation completes the task why would I make the success parameter false?
Return { success: true } when the operation is successful

Task:delete
- [ ] Return { success: true } when the operation is successfull

Task:getAll
- [ ] Missing pagination
- [ ] From does not seem to be a required parameter. Add It under the collection filter fields
- [ ] Rename filter fields collection to filters
- [ ] Remove Last ID parameter.
- [ ] It does not seem to return the task I created

Task:get
- [ ] Is the short id parameter really needed? Can just detect that behind the scenes and make the corresponding call?


Container resource
- [ ] I would probably only leave the get operation for this resource. Then, move insert and update to another resource called containerTask.

Container:insertTasks

- [ ] I would create different operations depending on the insert type. for example: containerTask:append, containerTask:Prepend.
- [ ] Rename tasks to Task IDs

Hub resource

Hub:getAll
- [ ] Missing pagination

Hub:update
- [ ] Destination should be under update fields

Team resource

Team:auto-dispatch

- [ ] What is road end exactly? an address?
- [ ] Return { success: true } when the operation is successfull

Team:create

- [ ] All the fields marked as required, are not actually required. Everything that is not required should not be at the main level but instead should be moved underneath a collection. In this case, the additional fields collection

Team:delete
- [ ] Return { success: true } when the operation is successfull

Team:getAll

- [ ] Missing pagination

Team:getTimeStimates

- [ ] Dropoff and pickup do not seem to be required parameters. Both need to be under additional fields if this is the case

Webhook resource

- [ ] Is it the webhook resource usufull in the node? I would not say so, since we are already using those endpoints in the trigger. Do you have a use case in mind where you need to create Webhooks dynamically? 

Worker resource

Worker:create
- [ ] It does not seem to be working. No matter the input I provide.

worker:getAll
- [ ] Missing pagination

**Comment by jamesliupenn** [comment]
**Admin resource**
Admin:getAll

- [ ] Admin:getall missing pagination

Admin:delete

- [ ] Return { success: true } when the operation is successful

**Task resource**

- [ ]  Do only task:create and do the batching behind the scenes. Batching if available should be the implementation to use.

Task:create
- [ ] Every field that is not required should be under a collection. I was able to create a task only with the destination address, which means, that recipients it’s not required
- [ ]  When setting the address you have a collection called “additional fields” that is not necessary. Add them at the same level of the destination address and make them not required. Do the same with Recipient
- [ ] Rename additional task fields to Additional Fields

Task:clone
- [ ]  If the operation does not return anything when it ran successfully return { success: true }
- [ ]  Rename collection options to additional fields
- [ ]  Move what is inside the collection override to the same level the options parameters are

Task:complete
- [ ]  If this operation completes the task why would I make the success parameter false?
- [ ] Return { success: true } when the operation is successful

Task:delete
- [ ]   Return { success: true } when the operation is successful

Task:getAll
- [ ]  Missing pagination
- [ ]  From does not seem to be a required parameter. Add It under the collection filter fields
- [ ] Rename filter fields collection to filters
- [ ]  Remove Last ID parameter.
- [ ]  It does not seem to return the task I created

Task:get
- [ ]  Is the short id parameter really needed? Can just detect that behind the scenes and make the corresponding call?

**Container resource**
- [ ]  I would probably only leave the get operation for this resource. Then, move insert and update to another resource called containerTask.

Container:insertTasks
- [ ]  I would create different operations depending on the insert type. for example: containerTask:append, containerTask:Prepend.
- [ ]  Rename tasks to Task IDs

**Hub resource**
Hub:getAll
- [ ] Missing pagination

Hub:update
- [ ]   Destination should be under update fields

**Team resource**
Team:auto-dispatch
- [ ]  What is road end exactly? an address?
- [ ]  Return { success: true } when the operation is successful

Team:create
- [ ]  All the fields marked as required, are not actually required. Everything that is not required should not be at the main level but instead should be moved underneath a collection. In this case, the additional fields collection

Team:delete
- [ ] Return { success: true } when the operation is successful

Team:getAll
- [ ]   Missing pagination

Team:getTimeStimates
- [ ]  Drop Off and Pick Up do not seem to be required parameters. Both need to be under additional fields if this is the case

**Webhook resource**
- [ ] Is it the Webhook resource useful in the node? I would not say so, since we are already using those endpoints in the trigger. Do you have a use case in mind where you need to create Webhooks dynamically?

**Worker resource**
Worker:create
- [ ]  It does not seem to be working. No matter the input I provide.

worker:getAll
- [ ]   Missing pagination

**Comment by jamesliupenn** [comment]
@YokySantiago to finish up the checklist items, @RicardoE105 this is over to you for review

**Comment by YokySantiago** [comment]
In order to complete what James mentioned

 **Task:getAll**

- [ ] It does not seem to return the task I created. Are you testing with an Onfleet account? Did you set a From value? From field is required for this operation but n8n it's not displaying it as required so, we created a new topic in n8n and in the meantime we added a manual validation. [n8n report](https://community.n8n.io/t/datetime-field-marked-as-required-is-not-working/11383)

**Container resource**
In this case we don't have container information. This is the meaning of container for us "Containers are an abstraction which describes task assignment. A container is an ordered list of tasks. Organizations, teams and workers all correspond to containers as they can all be assigned tasks." So, when we get information about a container, we need to select the type (Organization, Team, Worker) in order to get the task assignment for the container selected.
So, We would like to keep it all together as a resource

**Team:auto-dispatch**

- [ ]  What is road end exactly? an address?. This field is a select now with some additional option, improving the usability. **FIXED**

**Team:create**

- [ ] All the fields marked as required, are not actually required. Everything that is not required should not be at the main level but instead should be moved underneath a collection. In this case, the additional fields collection. **FIXED**

**Comment by jamesliupenn** [comment]
> @jamesliupenn @YokySantiago I'm trying to test the team resource, but it says "payment is required". Can you provide a test account with access to the team endpoints?
> 
> image
> 
> Also, to create a container I need a container ID. Where can I get a container ID from? 

What endpoint on the teams endpoint are you trying to access? If you are hitting specific endpoints like auto dispatch, you will need that feature enabled on your account and I can help you with that.

Containers are abstractions of task assignments, so if you assign a task to a worker/org/team, there will be a container created. The container IDs are be available via the worker/org/team ID, so basically depending on which entity you are accessing the container for, use that entity ID.

An example would be accessing a team's container, simply use that team ID

**Comment by jamesliupenn** [comment]
@RicardoE105 you should be on a test account and has the auto-dispatch endpoint enabled now

**Comment by janober** [comment]
Thanks a lot for your contribution. Got merged with #2845

**Comment by janober** [comment]
Got released with `n8n@0.165.0`

**Pull request closed without merging**